Add Day1 and Day2 solutions with tests

ya55en opened the pull request:
Add Day1 and Day2 solutions with tests.

ya55en committed:
Add Day1 and Day2 solutions with tests

ya55en committed:
Fix typos in README.md

ya55en committed:
Refactor a bit Day1 solution (renames)

ya55en committed:
Refactor Day2 solution (mostly renames)

emil-rb commented on src/day1/array_ext.rb:
```diff
@@ -0,0 +1,25 @@
+#: Extend Array with scan_zzz methods (see method docs).
+
+class Array
```
Although extending Array is possible, it's considered a bad practice for general code. It should be done only when the added method is used extensively (i.e. in Rails).

ya55en replied:
Got it, thanks, fixed in 5c87212.

emil-rb commented on src/day1/expense_report_fix.rb:
```diff
@@ -0,0 +1,83 @@
+
+module ExpenseReportFix
+
+  def ExpenseReportFix.scan_pairs(amounts, expected_sum = 2020)
```
When defining module-level methods the idiom is to do `extend self` in the beginning.
https://www.bigbinary.com/blog/extend-self-in-ruby

ya55en replied:
Thanks, `extend self` is a nice thing! (I don't think I've seen it in any of the books, btw). This one fixed in 5c87212; also reworked `ScanPasswds` and `AocUtils` the same way.

emil-rb commented on src/day2/scan_passwds.rb:
```diff
@@ -0,0 +1,95 @@
+
+  #: Return true if given passwd matches Toboggan Corporate Policies constraints,
+  #: false otherwise.
+  def ScanPasswds.match_constraints_tcp?(fst_idx, sec_idx, char, passwd)
```
Don't abbreviate variables. We consider this a bad practice.

ya55en replied:
Assuming you mean the arguments to `match_constraints_tcp?` - fixed in df1b427.

emil-rb commented on src/day2/scan_passwds.rb:
```diff
@@ -0,0 +1,95 @@
+    min_s, max_s, char, passwd = match_data.captures
+    min, max = min_s.to_i, max_s.to_i
+    # print "min=#{min}, max=#{max}, char=#{char}, passwd=#{passwd}\n"
+    match_method = ScanPasswds.method(match_method_sym)
```
Clever! But we don't like clever code :) An `if` would've been sufficient.

ya55en replied:
Indeed. Fixed in df1b427. Also fixed command line options propagation in `ExpenseReportFix` the same way.

emil-rb commented on src/day2/scan_passwds.rb:
```diff
@@ -0,0 +1,95 @@
+  def ScanPasswds.count_valid_passwords(input_file, match_method_sym)
+    count = 0
+    file_obj = input_file.respond_to?(:read) ? input_file : File.open(input_file)
+    file_obj.each_line do |line|
```
For bonus points you can do this in a functional way with `reduce` :)

ya55en replied:
Done with 8c5eadb :)

emil-rb commented on src/utils.rb:
```diff
@@ -0,0 +1,15 @@
+  def AocUtils.file_to_array(filename)
+    res = []
+    File.open(filename) do |f|
+      f.each_line do |line|
```
You can do `f.each_line.map { |l| l.strip.to_i }`

ya55en replied:
Yep, much more elegant. Done in 6b32ef8.

ya55en committed:
Refactor further Day 1 solution

 - Move scan_pairs() and scan_triads() out of Array and into
   ExspenseReportFix;

 - Apply 'extend self' idiom to ExspenseReportFix module;
 - Replace "clever" method invocation with an simple 'if';
 - Refactor unit tests.

ya55en committed:
Refactor further Day 2 solution

 - Rename argumets of match_constraints_tcp?();
 - Replace "clever" method invocation in passwd_valid?
   with a simple 'if';
 - Cosmetic improvements;
 - Refactor unit tests.

ya55en committed:
Refactor + rename AocUtils.file_to_array

ya55en committed:
Provide reduce-based impl of count_valid_passwords()

emil-rb commented on src/day2/scan_passwds.rb:
```diff
@@ -0,0 +1,100 @@
+
+  #: Return true if given passwd matches original shopkeeper's policies
+  #: constraints, false otherwise. (See day2/problems.txt for details.)
+  def match_constraints_orig?(min, max, char, passwd)
```
I meant to not abbreviate the names of the variables: `passwd` -> `password`. Min, max and char are fine. Those are widespread enough.

ya55en replied:
I see. fixed in 44e54b6.

emil-rb commented on src/day2/scan_passwds.rb:
```diff
@@ -0,0 +1,100 @@
+  #: constraints, false otherwise. (Note that TCP policy authors count from 1.
+  #: See day2/problems.txt for details.)
+  def match_constraints_tcp?(first, second, char, passwd)
+    (passwd[first - 1] == char) ^ (passwd[second - 1] == char)
```
What does this `^` do in this case?

ya55en replied:
Xor. This is an implementation of the "Toboggan Corporate Policy" for authentication, namely:
> Exactly one of these positions must contain the given letter.

Thus `match_constraints_tcp?` returns `true` iff exactly one of the positions contains the required character.

emil-rb replied:
OK, cool. Now I learned this operator can be used for booleans as well as bit math.

emil-rb commented on tests/day1-tests/test_expense_report_fix.rb:
```diff
@@ -0,0 +1,112 @@
+
+    def test_scan_triads__size_4
+      res = []
+      ExpenseReportFix.scan_triad_combinations(4) { |i, j, k| res << [i, j, k] }
```
You can fix this need to pass a block by doing this: https://stackoverflow.com/a/66059990/135665

ya55en replied:
Yep, nice, thank you! Done in e07ebc5.

ya55en committed:
Return enum_for __method__ from scan_pair/triad_combinations

 (Facilitating convertion to array when block not given.)

ya55en committed:
Rename 'passwd' (any case) to password
